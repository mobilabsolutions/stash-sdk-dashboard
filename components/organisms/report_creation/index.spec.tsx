import ReportCreation from './index'
import { testRender, deepRender } from '../../../test_utils'
import { fireEvent, wait } from 'react-testing-library'

jest.mock('moment', () => {
  const moment = require.requireActual('moment')
  const mocked = () => moment('20190916')
  mocked.defineLocale = jest.fn()
  mocked.locale = jest.fn()
  mocked.isMoment = jest.fn()
  mocked.localeData = () => ({
    longDateFormat: jest.fn()
  })
  return mocked
})

describe('Report creation', () => {
  test('should list default reports', () => {
    testRender(ReportCreation, {
      show: true,
      onClose: jest.fn()
    })
  })
  test('should trigger dowload', () => {
    const downloadReport = jest.fn()
    const { getByText } = deepRender(ReportCreation, {
      show: true,
      downloadReport,
      onClose: jest.fn()
    })
    getByText('Monthly Transactions Overview').click()
    expect(downloadReport).toHaveBeenCalledWith(
      'Monthly Transactions Overview_20190916',
      'default',
      { reportType: 'OVERVIEW' }
    )
    getByText('Chargebacks Overview').click()
    expect(downloadReport).toHaveBeenCalledWith(
      'Chargebacks Overview_20190916',
      'default',
      { reportType: 'CHARGEBACK' }
    )
    getByText('Monthly Refunds Overview').click()
    expect(downloadReport).toHaveBeenCalledWith(
      'Monthly Refunds Overview_20190916',
      'default',
      { reportType: 'REFUND' }
    )
  })
})

describe('Custom reports creation', () => {
  test('should show filters for custom report', () => {
    const downloadReport = jest.fn()
    const { getByText, baseElement } = deepRender(ReportCreation, {
      show: true,
      downloadReport,
      onClose: jest.fn()
    })
    getByText('Set up your own report').click()
    expect(baseElement.firstChild).toMatchSnapshot()
  })

  test('should expand filters for custom reports', () => {
    const downloadReport = jest.fn()
    const { getByText, baseElement } = deepRender(ReportCreation, {
      show: true,
      downloadReport,
      onClose: jest.fn()
    })
    getByText('Set up your own report').click()
    getByText('Show more').click()
    getByText('Show less')
    expect(baseElement.firstChild).toMatchSnapshot()
  })

  test('should disable if not Title and one filter', async () => {
    const downloadReport = jest.fn()
    const { getByText, getByPlaceholderText } = deepRender(ReportCreation, {
      show: true,
      downloadReport,
      onClose: jest.fn()
    })
    getByText('Set up your own report').click()
    const GenerateBtn = getByText('Generate Report').parentElement

    expect(GenerateBtn.getAttribute('disabled')).toBeDefined()
    fireEvent.change(getByPlaceholderText('Report Name'), {
      target: { value: 'New Report' }
    })
    expect(GenerateBtn.getAttribute('disabled')).toBeDefined()
    fireEvent.change(getByPlaceholderText('Keyword'), {
      target: { value: 'new key word' }
    })
    await wait()
    expect(GenerateBtn.getAttribute('disabled')).toBe('')
  })

  // Skip until we update react and testing-library https://github.com/testing-library/react-testing-library/issues/281
  // it('should trigger download', async () => {
  //   jest.useFakeTimers()
  //   const downloadReport = jest.fn()
  //   const { getByText, getByPlaceholderText, baseElement } = deepRender(
  //     ReportCreation,
  //     {
  //       show: true,
  //       downloadReport,
  //       onClose: jest.fn()
  //     }
  //   )

  //   getByText('Set up your own report').click()

  //   fireEvent.change(getByPlaceholderText('Report Name'), {
  //     target: { value: 'New Report' }
  //   })
  //   //Open select menu typing on input
  //   fireEvent.change(
  //     queryByAttribute('id', baseElement, 'react-select-8-input'),
  //     {
  //       target: { value: 'a' }
  //     }
  //   )

  //   //Click first option
  //   fireEvent.click(
  //     queryByAttribute('id', baseElement, 'react-select-8-option-1')
  //   )
  //   getByText('Generate Report').parentElement.click()

  //   await wait()

  //   expect(downloadReport).toHaveBeenCalledWith('New Report', 'custom', {
  //     action: null,
  //     amount: '',
  //     createdAtEnd: null,
  //     createdAtStart: null,
  //     currency: '',
  //     customerId: '',
  //     merchantTransactionId: '',
  //     paymentMethod: 'CC',
  //     status: null,
  //     text: '',
  //     transactionId: ''
  //   })
  // })
})
