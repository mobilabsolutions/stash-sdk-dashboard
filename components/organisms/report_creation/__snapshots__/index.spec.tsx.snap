// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report creation should list default reports 1`] = `
<div
  className="overlay-sc-1up1mft-0 MmcNI"
  onClick={[Function]}
>
  <div
    className="popup__Content-br7oqf-0 dLDWKT"
  >
    <div
      className="action_popup__PopupContainer-sc-6dvik9-0 fEHenQ"
    >
      <div
        className="action_popup__HeaderContainer-sc-6dvik9-3 pQDGY"
      >
        <span
          className="action_popup__HeaderTitle-sc-6dvik9-4 cmNKLI"
        >
          Select one of the default reports
        </span>
        <div
          className="action_popup__HeaderClose-sc-6dvik9-5 fSRmOH"
        >
          <svg
            height={24}
            onClick={[MockFunction]}
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.707 11.293l5.657-5.657-5.657 5.657zM12 12zm-.707.707l-5.657 5.657 5.657-5.657zM13.414 12l5.657 5.657a1 1 0 0 1-1.414 1.414L12 13.414l-5.657 5.657a1 1 0 1 1-1.414-1.414L10.586 12 4.929 6.343A1 1 0 0 1 6.343 4.93L12 10.586l5.657-5.657a1 1 0 0 1 1.414 1.414L13.414 12zm-.707.707l5.657 5.657-5.657-5.657zm-1.414-1.414L5.636 5.636l5.657 5.657z"
              fill="#a3aaaf"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        className="report_creation__Wrapper-sc-1bzzh01-1 idsckk"
      >
        <div
          className="report_creation__ItemWrapper-sc-1bzzh01-0 cMHPBJ"
          onClick={[Function]}
        >
          <div
            className="check-mark"
          >
            <svg
              height={16}
              viewBox="0 0 40 40"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="20"
                  cy="20"
                  fill="#d1d5d7"
                  r="20"
                />
                <path
                  d="M11.667 21.481l5.128 5.186L28.333 15"
                  stroke="#FFF"
                  strokeLinecap="round"
                  strokeWidth="2"
                />
              </g>
            </svg>
          </div>
          <div
            className="central-icon"
          >
            <svg
              height={32}
              viewBox="0 0 32 32"
              width={32}
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="16"
                  cy="16"
                  fill="#B7D2FB"
                  r="16"
                />
                <path
                  d="M21.672 12.719c.219.219.328.479.328.781v9.375c0 .313-.11.578-.328.797a1.085 1.085 0 0 1-.797.328h-9.75c-.313 0-.578-.11-.797-.328a1.085 1.085 0 0 1-.328-.797V9.125c0-.313.11-.578.328-.797.219-.219.484-.328.797-.328h5.406c.157 0 .308.031.453.094.021.01.037.02.047.031.021 0 .042.01.063.031.094.052.172.11.234.172l4.344 4.39zm-1.422.156l-3.125-3.156v3.156h3.125zm.625 10V14h-3.75c-.313 0-.578-.11-.797-.328a1.085 1.085 0 0 1-.328-.797v-3.75h-4.875v13.75h9.75z"
                  fill="#2E7EF2"
                />
              </g>
            </svg>
          </div>
          <span
            className="title"
          >
            Transactions Monthly Overview
          </span>
        </div>
        <div
          className="report_creation__ItemWrapper-sc-1bzzh01-0 cMHPBJ"
          onClick={[Function]}
        >
          <div
            className="check-mark"
          >
            <svg
              height={16}
              viewBox="0 0 40 40"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="20"
                  cy="20"
                  fill="#d1d5d7"
                  r="20"
                />
                <path
                  d="M11.667 21.481l5.128 5.186L28.333 15"
                  stroke="#FFF"
                  strokeLinecap="round"
                  strokeWidth="2"
                />
              </g>
            </svg>
          </div>
          <div
            className="central-icon"
          >
            <svg
              height={32}
              viewBox="0 0 32 32"
              width={32}
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="16"
                  cy="16"
                  fill="#B7D2FB"
                  r="16"
                />
                <path
                  d="M21.672 12.719c.219.219.328.479.328.781v9.375c0 .313-.11.578-.328.797a1.085 1.085 0 0 1-.797.328h-9.75c-.313 0-.578-.11-.797-.328a1.085 1.085 0 0 1-.328-.797V9.125c0-.313.11-.578.328-.797.219-.219.484-.328.797-.328h5.406c.157 0 .308.031.453.094.021.01.037.02.047.031.021 0 .042.01.063.031.094.052.172.11.234.172l4.344 4.39zm-1.422.156l-3.125-3.156v3.156h3.125zm.625 10V14h-3.75c-.313 0-.578-.11-.797-.328a1.085 1.085 0 0 1-.328-.797v-3.75h-4.875v13.75h9.75z"
                  fill="#2E7EF2"
                />
              </g>
            </svg>
          </div>
          <span
            className="title"
          >
            Refunds Monthly Overview
          </span>
        </div>
        <div
          className="report_creation__ItemWrapper-sc-1bzzh01-0 cMHPBJ"
          onClick={[Function]}
        >
          <div
            className="check-mark"
          >
            <svg
              height={16}
              viewBox="0 0 40 40"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="20"
                  cy="20"
                  fill="#d1d5d7"
                  r="20"
                />
                <path
                  d="M11.667 21.481l5.128 5.186L28.333 15"
                  stroke="#FFF"
                  strokeLinecap="round"
                  strokeWidth="2"
                />
              </g>
            </svg>
          </div>
          <div
            className="central-icon"
          >
            <svg
              height={32}
              viewBox="0 0 32 32"
              width={32}
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="16"
                  cy="16"
                  fill="#B7D2FB"
                  r="16"
                />
                <path
                  d="M21.672 12.719c.219.219.328.479.328.781v9.375c0 .313-.11.578-.328.797a1.085 1.085 0 0 1-.797.328h-9.75c-.313 0-.578-.11-.797-.328a1.085 1.085 0 0 1-.328-.797V9.125c0-.313.11-.578.328-.797.219-.219.484-.328.797-.328h5.406c.157 0 .308.031.453.094.021.01.037.02.047.031.021 0 .042.01.063.031.094.052.172.11.234.172l4.344 4.39zm-1.422.156l-3.125-3.156v3.156h3.125zm.625 10V14h-3.75c-.313 0-.578-.11-.797-.328a1.085 1.085 0 0 1-.328-.797v-3.75h-4.875v13.75h9.75z"
                  fill="#2E7EF2"
                />
              </g>
            </svg>
          </div>
          <span
            className="title"
          >
            Chargebacks
          </span>
        </div>
        <div
          className="report_creation__ItemWrapper-sc-1bzzh01-0 dKMRZh"
          onClick={[Function]}
        >
          <div
            className="check-mark"
          />
          <div
            className="central-icon"
          >
            <svg
              height="32"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
              >
                <circle
                  cx="16"
                  cy="16"
                  fill="#B7D2FB"
                  r="16"
                />
                <path
                  d="M20 16.586h-3.348V20h-1.32v-3.414H12v-1.188h3.331V12h1.321v3.398H20z"
                  fill="#2E7EF2"
                  fillRule="nonzero"
                />
              </g>
            </svg>
          </div>
          <span
            className="title"
          >
            Set up your own report
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
